lectures/opt_savings.md

```diff
@@ -0,0 +1,455 @@
+    while error > 0:
+        v_σ = get_value(σ, constants, sizes, arrays)
+        σ_new = get_greedy(v_σ, constants, sizes, arrays)
+        error = jnp.max(np.abs(σ_new - σ))
+        σ = σ_new
+        i = i + 1
+        print(f"Concluded loop {i} with error {error}.")
+    return σ
+```
+
+```{code-cell} ipython3
+def optimistic_policy_iteration(model, tol=1e-5, m=10):
+    "Implements the OPI routine."
+    constants, sizes, arrays = model
+    v = jnp.zeros(sizes)
+    error = tol + 1
+    while error > tol:
+        last_v = v
+        σ = get_greedy(v, constants, sizes, arrays)
+        for _ in range(m):
+            v = T_σ(v, σ, constants, sizes, arrays)
+        error = jnp.max(np.abs(v - last_v))
+    return get_greedy(v, constants, sizes, arrays)
+```
+
+## Plots
+
+Create a JAX model for consumption, perform policy iteration, and plot the resulting optimal policy function.
+
+```{code-cell} ipython3
+fontsize=12
+model = create_consumption_model_jax()
+# Unpack 
+constants, sizes, arrays = model
+β, R, γ = constants
+w_size, y_size = sizes
+w_grid, y_grid, Q = arrays
+σ_star = policy_iteration(model)
+fig, ax = plt.subplots(figsize=(9, 5.2))
+ax.plot(w_grid, w_grid, "k--", label="45")
+ax.plot(w_grid, w_grid[σ_star[:, 1]], label="$\\sigma^*(\cdot, y_1)$")
+ax.plot(w_grid, w_grid[σ_star[:, -1]], label="$\\sigma^*(\cdot, y_N)$")
+ax.legend(fontsize=fontsize)
+plt.show()
+```
+
+## Tests
+
+Here's a quick test of the timing of each solver.
+
+```{code-cell} ipython3
+model = create_consumption_model_jax()
+```
+
+```{code-cell} ipython3
+print("Starting HPI.")
+start_time = time.time()
+out = policy_iteration(model)
+elapsed = time.time() - start_time
+print(f"HPI completed in {elapsed} seconds.")
+```
+
+```{code-cell} ipython3
+print("Starting VFI.")
+start_time = time.time()
+out = value_iteration(model)
+elapsed = time.time() - start_time
+print(f"VFI(jax not in succ) completed in {elapsed} seconds.")
+```
+
+```{code-cell} ipython3
+print("Starting OPI.")
+start_time = time.time()
+out = optimistic_policy_iteration(model, m=100)
+elapsed = time.time() - start_time
+print(f"OPI completed in {elapsed} seconds.")
+```
+
+```{code-cell} ipython3
+def run_algorithm(algorithm, model, **kwargs):
+    start_time = time.time()
+    result = algorithm(model, **kwargs)
+    end_time = time.time()
+    elapsed_time = end_time - start_time
+    print(f"{algorithm.__name__} completed in {elapsed_time:.2f} seconds.")
+    return result, elapsed_time
+
+model = create_consumption_model_jax()
+σ_pi, pi_time = run_algorithm(policy_iteration, model)
+σ_vfi, vfi_time = run_algorithm(value_iteration, model, tol=1e-5)
+
+m_vals = range(5, 3000, 100)
+opi_times = []
+for m in m_vals:
+    σ_opi, opi_time = run_algorithm(optimistic_policy_iteration, model, m=m, tol=1e-5)
+    opi_times.append(opi_time)
+
+fig, ax = plt.subplots(figsize=(9, 5.2))
+ax.plot(m_vals, np.full(len(m_vals), pi_time), lw=2, label="Howard policy iteration")
```

Comment:
You can either use `jnp.full` or `[pi_time]* m_vals`. I don't think that would make much performance difference, but yeah that removes `numpy` completely.